### PR TITLE
Reference lines and reference areas

### PR DESCRIPTION
**Pull request by hughess**
### Description
This PR adds reference lines and areas to charts.

For now, these are separate components: `ReferenceLine` and `ReferenceArea`, which are used by inserting them into a slot within a chart. For example:

```html
<LineChart
   data={sales_data}
   x=date
   y=sales
>
   <ReferenceLine yVal=120 label="Target"/>
</LineChart>
```

These can be applied by hardcoding values (e.g., if you want a horizontal line at y=50) or with a dataset (e.g., you have a table of events or time ranges you want to include in a chart to add context to the viz).

#### Props

Required:
- `x` or `y`

Optional:
- `data`
- `label`
- `color`
- `lineType`
- `lineColor`
- `labelColor`
- `lineWidth`
- `showValueInLabel`
- `labelTextOutline`
- `labelBackground`

CleanShot 2023-05-07 at 12 26 31@2x

```html
<LineChart 
    data={daily_complaints} 
    x=date 
    y=number_of_complaints 
    title="Complaint Calls to Austin 311"
>
    <ReferenceLine yVal=2500 label=Threshold/>
</LineChart>
```

#### Props

Required:
- At least one of `x1`, `x2`, `y1`, `y2`

Optional:
- `data`
- `label`
- `color`
- `opacity`
- `labelColor`
- `border`
- `borderColor`
- `borderType`
- `borderWidth`
- `labelPosition`

CleanShot 2023-05-06 at 15 23 05@2x

```html
<LineChart 
    data={daily_complaints} 
    x=date 
    y=number_of_complaints 
    title="Complaint Calls to Austin 311"
>
    <ReferenceArea data={annotate} xMin=start_date xMax=end_date label=label/>
</LineChart>
```

### Use Cases
This opens up some pretty exciting use cases for adding relevant context to data viz.
CleanShot 2023-05-07 at 12 36 50@2x

## To Do
- ~~Add `lineType` prop to reference line (dashed, solid, etc)~~
- ~~Formatted label numbers~~
- ~~How to treat label when no label is supplied (show value or nothing?)~~
- ~~Fix gridline z level - histogram issue~~
- ~~Fix tooltip hover issue by setting markLine and markArea to silent~~
- ~~Work in composable charts?~~
- ~~Open slots on all XY chart types~~
- ~~Naming - highlight area?~~
- ~~Switch `yVal` to `y`?~~
- ~~Switch `yMin` to `y0` or `y1`?~~
- ~~Default label positions (swapXY, chart types, etc.)~~
- ~~Error handling~~
- ~~Styling for labels - add white outline to text? Inherit colour of area background for that outline?~~
- ~~Add docs~~
- ~~Add note to docs for gridline overlap issue with reference area label~~
- ~~Add note to docs about continuous axis bar charts - reference area doesn't cover bar by default~~

### Deferred for future upgrades
- Reference area should cover full bar width area rather than stopping at midpoint of bar. This is fixed in Echarts v5.4.2, so an upgrade of that dependency will be required. Unclear if this will require the mark series to be `bar` rather than `line` as it is now
   - This fix works only for categorical axes 
- Shorthand for using in normal chart props
- Colours - offer presets or a default opacity level?
   - It would be nice to be able to use `color=red` in `ReferenceArea` and have it produce a faded red/pink background with a darker red label 
- Label word breaking/overflow behaviour
- Control over value formatting in label
- Legend options for reference area

### Issues to be solved
- Label rotation for vertical reference lines (see echarts issue here: https://github.com/apache/echarts/issues/13153)
- For reference areas, label z level should be different than area z level
   - Area should be behind gridlines, but label should be in front of them
   - Open echarts issues for this here:  
      - https://github.com/apache/echarts/issues/18597
      - https://github.com/apache/echarts/issues/18608

### Error Handling
Implemented using the `ErrorChart` component.

Covers these scenarios:
CleanShot 2023-05-09 at 15 56 10@2x

### Checklist
- [x] For UI or styling changes, I have added a screenshot or gif showing before & after
- [x] I have added a [changeset](https://github.com/evidence-dev/evidence/blob/main/CONTRIBUTING.md#adding-a-changeset)

### CODE REVIEW

**Comment by hughess** [comment]
The last outstanding issue that may need to be tackled before merging is the default styling of labels. The basic labels look fine, but when a label is overtop of lines/bars/etc it doesn't look right.

Options are adding a white outline on the text or adding a background colour to the label (either full opacity or with some transparency).

Given the number of permutations here, it might be worth merging with a small white text outline as a starting point, and spending time to solve this in a more general sense once we have a bit of user feedback.

**Comment by hughess** [comment]
Update: I've added a white text outline by default on reference line labels, but have also included a prop for `labelBackground` which if set to true, will include a white semi-transparent background behind the label

File: sites/example-project/src/components/viz/ReferenceLine.svelte [hunk 1]
```diff
@@ -0,0 +1,162 @@
+<script>
+	import { getContext } from 'svelte';
+	import { propKey, configKey } from './context';
+	import { formatValue } from '$lib/modules/formatting.js';
+	import checkInputs from '$lib/modules/checkInputs';
+	import ErrorChart from './ErrorChart.svelte';
+
+	let props = getContext(propKey);
```
Comment (ItsMeBrianD) [inline sites/example-project/src/components/viz/ReferenceLine.svelte:8]: Ditto above

**Review by ItsMeBrianD — COMMENTED** [review]
Nothing breaking; this looks super cool

Couple of nitpicky comments that might be worth looking at; otherwise LGTM

**Review by archiewood — COMMENTED** [review]
This is me naively trying stuff in order, and resorting to the docs when I have to. I think it's instructive

## ReferenceArea

```

<LineChart
    title="If the line is outside the default y range"
    data={nps}>
    <ReferenceLine y=70/>
</LineChart>


<LineChart
    title="Manually specify the range"
    yMin=-100
    yMax=100
    data={nps}>
    <ReferenceLine y=70/>
</LineChart>

<LineChart
    title="this doesnt look right"
    yMin=-100
    yMax=100
    data={nps}>
    <ReferenceArea yMin=70/>
</LineChart>

<LineChart
    title="this also doesnt work"
    yMin=-100
    yMax=100
    data={nps}>
    <ReferenceArea y0=70 y1=100/>
</LineChart>

<LineChart
    title="okay now i read the docs"
    yMin=-100
    yMax=100
    data={nps}>
    <ReferenceArea y1=70 y2=100/>
    
</LineChart>


<LineChart
    title="css colours are horrible"
    yMin=-100
    yMax=100
    data={nps}>
    <ReferenceArea y1=70 y2=100 color=green/>
    <ReferenceArea y1=30 y2=70 color=orange/>
    <ReferenceArea y2=30 color=red/>
</LineChart>


<LineChart
    title="fixed colors"
    yMin=-100
    yMax=100
    data={nps}>
    <!-- lightgreen -->
    <ReferenceArea y1=70 y2=100 color=#d8f2d8 label=Good/>
    <ReferenceArea y1=30 y2=70 color=#fffcd4 label=Okay/>
    <ReferenceArea y2=30 color=#ffeceb label= Bad/>
</LineChart>

<LineChart
    title="fixed"
    yMin=-100
    yMax=100
    data={nps}>
    <!-- lightgreen -->
    <ReferenceArea y1=70 y2=100 color=#d8f2d8 label=Good labelPosition=right/>
    <ReferenceArea y1=30 y2=70 color=#fffcd4 label=Okay labelPosition=right/>
    <ReferenceArea y2=30 color=#ffeceb label= Bad labelPosition=right/>
</LineChart>
```


image


## ReferenceLine

```
<BarChart     
    data={orders_by_day}
    title=works>
    <ReferenceLine y=4000/>
</BarChart>

<BarChart 
    data={orders_by_day}
    title="wrong prop name">
    <ReferenceLine y=4000 name ="Q3 Forecast"/>
    <ReferenceLine y=3800 name ="RF2"/>
</BarChart>


<BarChart 
    data={orders_by_day}
    title="right prop name, overlapping labels">
    <ReferenceLine y=4000 label ="Q3 Forecast"/>
    <ReferenceLine y=3800 label ="RF2"/>
</BarChart>


<BarChart 
    data={orders_by_day}
    title="read the docs, how do I use labelPosition">
    <ReferenceLine y=4000 label ="Q3 Forecast"/>
    <ReferenceLine y=3800 label ="RF2" labelPosition=bottom/>
</BarChart>

<BarChart 
    data={orders_by_day}
    title="still dont get it">
    <ReferenceLine y=4000 label ="Q3 Forecast"/>
    <ReferenceLine y=3800 label ="RF2" labelPosition=bottomRight/>
</BarChart>

<BarChart 
    data={orders_by_day}
    title="color">
    <ReferenceLine y=4000 label ="Q3 Forecast"/>
    <ReferenceLine y=3800 label ="RF2" color=green/>
</BarChart>

<BarChart 
    data={orders_by_day}
    title="I really just want to hide the values when I have labels">
    <ReferenceLine y=4000 label ="Q3 Forecast" hideValue/>
    <ReferenceLine y=3800 label ="RF2" color=green hideValue/>
</BarChart>

<BarChart 
    data={orders_by_day}
    title="or format them">
    <ReferenceLine y={fmt("$###", 4000)} label ="Q3 Forecast" hideValue/>
    <ReferenceLine y=3800 label ="RF2" color=green hideValue/>
</BarChart>

```
image


Overall thoughts:
- RefArea: I can get the result I want, but takes more hacking than I expected
- RefLine: I cant get the result I want
- PropNames: 
   - `y1` and `y2` was not that intuitive
   - `hideValue` was my expected name for hideValueInLabel
- I already want a shorthand to get the ReferenceLine as a prop in the barchart (maybe only if I don't configure it at all except a value. 
- I think I want more precisions on target line label than the axis. Should I be able to configure this?
- Area: What's the right default label position? I would guess right[Center]
- Line: Should it be 2px by default not 1px? (or just thicker)

File: sites/example-project/src/components/viz/ReferenceArea.svelte [hunk 2]
```diff
@@ -0,0 +1,170 @@
+<script>
+	import { getContext } from 'svelte';
+	import { propKey, configKey } from './context';
+	import checkInputs from '$lib/modules/checkInputs';
+	import ErrorChart from './ErrorChart.svelte';
+	let props = getContext(propKey);
```
[the remark this answers is not in the document]
Comment (hughess) [reply sites/example-project/src/components/viz/ReferenceArea.svelte:6]: I pulled this from `Line.svelte` for consistency. Do you think it's worth changing, or leave for now and change across the charting library in the future?

File: sites/example-project/src/components/viz/ReferenceArea.svelte [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		case 'left':
+			labelPosition = 'insideLeft';
+			break;
+		case 'center':
+			labelPosition = 'inside';
+			break;
+		case 'centre':
+			labelPosition = 'inside';
+			break;
+	}
```
[the remark this answers is not in the document]
Comment (hughess) [reply sites/example-project/src/components/viz/ReferenceArea.svelte]: Sounds good - I'll make that change!

**Comment by hughess** [comment]
Thanks @archiewood. 

- What were you trying to achieve with reference line?
   - I recently added in `labelPosition` for reference line, so that might help some of the issues there
- Agree on the prop names. Think I'll switch back to `yMin` and `yMax` for ref area
- Preset colours are now included to override the CSS colour names. It includes:
   - red
   - green
   - yellow
   - blue
   - grey
- I've set the default area label position to top left, but mostly because of the x-axis case (e.g., date range). I agree if it's y-based, it should probably be right centre   
- Good idea on `hideValue`
- Agreed on shorthand within the component. I think better to tackle that in the next version - there might be a way to get reference area into that strategy as well
- Line might need to be thicker - especially when there's a lot of data on a chart
   - Unrelated, but I'm also wondering if our bars are too spaced out on the bar chart - looks visually overwhelming. Might also just be the high contrast between the blue and the white background

> I think I want more precisions on target line label than the axis. Should I be able to configure this?

What do you mean by precisions?

**Comment by archiewood** [comment]
I was trying to achieve two lines where the labels didn't overlap

By more _precision_ (typo), here I meant I wanted the format to show $3,800 not $4k.

**Comment by hughess** [comment]
Ahh that's a great point. Do you think a good solution to that might be a format override prop like `valueFmt` or similar?

**Comment by archiewood** [comment]
Yes I think so. 
Or accept the format function as I tried and failed to do

**Comment by hughess** [comment]
I've added value format control to the list of future improvements.

Just pushed some simplification to the props and label positions

**Comment by archiewood** [comment]
@hughess as discussed I think we should probably leave the behaviour as is for numeric axes highlighting.

It's kind of a difficult problem as barcharts are sort of a discrete chart type, but are often used in a continuous setting (eg dates, numeric ranges). 

Given the workarounds exist (workaround 1 in particular), I think until we are clear on what users need here, leaving as is makes sense

**Comment by hughess** [comment]
There are only 2 things blocking this PR from merging:
1. Format test failing, but this is only for markdown pages - @ItsMeBrianD wanted to double check that this is okay
2. Docs - since this includes docs changes, assume we should wait until we're close to release?

**Comment by archiewood** [comment]
Removed formatting checks for markdown files in the example project `+page.md` files.

**Pull request merged**